lib/spack/docs/packaging_guide.rst

```diff
@@ -2775,6 +2775,254 @@ Suppose a user invokes ``spack install`` like this:
 Spack will fail with a constraint violation, because the version of
 MPICH requested is too low for the ``mpi`` requirement in ``foo``.
 
+.. _custom-attributes:
+
+------------------
+Custom attributes
+------------------
+
+Often packages will need to provide attributes for dependents to query
+various details about what it provides.  While any number of custom defined
+attributes can be implemented by a pacakage, the specific four attributes
+described below are always available on every package with default
+implementations and the ability to customize with alternate implementations
+in the case of virtual packages provided:
+
+=========== =========================================== =====================
+Attribute   Purpose                                     Default
+=========== =========================================== =====================
+``root``    The installation directory for the package  ``spec.prefix``
+``command`` An executable command for the package       | ``spec.name`` in
+                                                        | ``spec.prefix.bin``
+``headers`` A list of headers provided by the package   | All headers
+                                                          searched
+                                                        | recursively in
+                                                          ``spec.prefix``
+``libs``    A list of libraries provided by the package | ``lib{spec.name}``
+                                                          searched
+                                                        | recursively in
+                                                          ``lib``, ``lib64``,
+                                                        | then the rest of
+                                                          of ``spec.prefix``
```

Comment:
why not write the `headers` and `libs` defaults in terms of `spec.root`, so that if you change the root for a virtual, you automatically get appropriately implemented `headers` and `libs` attributes too?  I think this behavior will be more common, and you could still override both `foo_root` and `foo_headers` if you need to, but more likely you'll only need to override `foo_root`.

Reply:
This got me stuck for quite a while.  I agree it should be implemented this way but there's something about the `libs` property that completely breaks down when you try to call another custom property inside it's implementation.  The problem is unrelated to this change as I can reproduce it in current `develop` so I'll leave it as a TODO afterwards.